table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/CompatibleMetastoreClient.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive.util;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.thrift.TApplicationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import javax.annotation.Nullable;
+
+class CompatibleMetastoreClient implements InvocationHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(CompatibleMetastoreClient.class);
+
+  private final IMetaStoreClient mDelegate;
+  private final HMSShim mCompat;
+
+  CompatibleMetastoreClient(IMetaStoreClient delegate, @Nullable HMSShim compatibility) {
+    mDelegate = delegate;
+    mCompat = compatibility;
+  }
+
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    try {
+      return method.invoke(mDelegate, args);
+    } catch (InvocationTargetException delegateException) {
+      try {
+        if (mCompat != null
+            && delegateException.getCause().getClass()
+            .isAssignableFrom(TApplicationException.class)) {
+          LOG.debug("Attempting to call hive metastore with {}", mCompat.getClass().getName());
```

Comment:
Does the name of the method already imply the compatibility shim? If not, can we be explicit in the message that we are using the compatibility shim?

Reply:
done